pkg/remotecmd/kubeexec.go

```diff
@@ -0,0 +1,294 @@
+		}
+
+		if outputHandler != nil {
+			outputHandler(Stopped, stdout, stderr, err)
+		}
+	}()
+
+	return nil
+}
+
+// StopProcessForCommand stops the process representing the specified Devfile command.
+// Because of the way this process is launched and its PID stored (see StartProcessForCommand),
+// we need to determine the process children (there should be only one child). Then killing those children
+// will exit the parent 'sh' process.
+func (k *kubeExecProcessHandler) StopProcessForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+) error {
+	klog.V(4).Infof("StopProcessForCommand for %q", devfileCmd.Id)
+	defer func() {
+		pidFile := getPidFileForCommand(devfileCmd)
+		err := ExecuteCommand([]string{common.ShellExecutable, "-c", fmt.Sprintf("rm -f %s", pidFile)},
+			kclient, podName, containerName, false, nil, nil)
+		if err != nil {
+			klog.V(2).Infof("Could not remove file %q: %v", pidFile, err)
+		}
+	}()
+
+	ppid, err := getRemoteProcessPID(kclient, devfileCmd, podName, containerName)
+	if err != nil {
+		return err
+	}
+	if ppid == 0 {
+		return nil
+	}
+
+	children, err := getProcessChildren(ppid, kclient, podName, containerName)
+	if err != nil {
+		return err
+	}
+
+	kill := func(p int) error {
+		err = ExecuteCommand([]string{common.ShellExecutable, "-c", fmt.Sprintf("kill %d || true", p)},
+			kclient, podName, containerName, false, nil, nil)
+
+		//Because the process we just stopped might take longer to exit (it might have caught the signal and is performing additional cleanup),
+		//retry detecting its actual state till it is stopped or timeout expires
+		var processInfo interface{}
+		processInfo, err = task.NewRetryable(fmt.Sprintf("status for remote process %d", p), func() (bool, interface{}, error) {
+			pInfo, e := k.getProcessInfoFromPid(p, kclient, podName, containerName)
+			return e == nil || pInfo.Status == Stopped, pInfo, e
+		}, true).RetryWithSchedule(2*time.Second, 4*time.Second, 8*time.Second)
+		if err != nil {
+			return err
+		}
+
+		pInfo, ok := processInfo.(RemoteProcessInfo)
+		if !ok {
+			klog.V(2).Infof("invalid type for remote process (%d) info, expected RemoteProcessInfo", p)
+			return fmt.Errorf("internal error while checking remote process status: %d", p)
+		}
+		if pInfo.Status != Stopped {
+			return fmt.Errorf("invalid status for remote process %d: %+v", p, processInfo)
+		}
+		return nil
+	}
+
+	if len(children) == 0 {
+		return kill(ppid)
+	}
+
+	for _, child := range children {
+		if err = kill(child); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getRemoteProcessPID(kclient kclient.ClientInterface, devfileCmd devfilev1.Command, podName string, containerName string) (int, error) {
+	pidFile := getPidFileForCommand(devfileCmd)
+	stdout, stderr, err := ExecuteCommandAndGetOutput(kclient, podName, containerName, false,
+		common.ShellExecutable, "-c", fmt.Sprintf("cat %s || true", pidFile))
```

Comment:
Yes, it was in case the pid file does not exist, for example the very first time we check if the process is running. I didn't want the command to fail (regardless of the error), because it would make it a bit complicated (for nothing, IMO, as this is not an issue) to additionally check `err` for the type of exit code. What matters most here is the content of the standard output of this `cat` command.